File: scripts/gpt/gpt-abstract-screen.py

import os
import json
import pickle
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)  # Display any number of columns
pd.set_option('display.expand_frame_repr', False)  # Do not wrap to multiple pages

import openai
import time
from datetime import datetime
# import requests
# from tenacity import retry, wait_random_exponential, stop_after_attempt

# set variables
GPT_MODEL = "gpt-3.5-turbo-0613"
openai.api_key = os.getenv('OPENAI_API_KEY')  # set openai api key, stored via Pycharm settings "Build, Execution, Deployment"
cost_per_1000_input_tokens = 0.0015  # $0.0015 per 1000 input tokens, https://openai.com/pricing 2023-06-15 dollars
cost_per_1000_output_tokens = 0.002

# TODO:

# FIXME:

# ==========================
# Custom GPT Functions
# ==========================

### questions for the model and model function
def my_gpt_api_function(abstract_disease: str, mirna_oi: str):
    """
    Create a json schema
    https://community.openai.com/t/emulated-multi-function-calls-within-one-request/269582

    :param abstract_disease: The disease as a string
    :param mirna_oi: The microRNA of interest in the abstract
    """
    functions = [
        {
            "name": "mirco_rna_abstract_data",
            "description": "Retrieve study data from abstract",
            "parameters": {
                "type": "object",
                "properties": {
                    # Q1
                    "related_topic": {
                        "type": "string",
                        "enum": ["Yes", "No", "Not Sure"],
                        "description": f"Does the abstract deal with {abstract_disease} in humans?",
                    },
                    # Q2
                    "direction_upreg_downreg": {
                        "type": "string",
                        "enum": ["Upregulated", "Downregulated", "Not Given"],
                        "description": f"Is the given miRNA of interest {mirna_oi} up- (increased) or downregulated (decreased) in {abstract_disease}?",
                    },
                    # Q3
                    "primary_literature": {
                        "type": "string",
                        "enum": ["Yes", "No", "Not Sure"],
                        "description": "Is the abstract from an original research (no review, meta-analyses)?",
                    },
                    # Q4
                    "serum_plasma_tissue": {
                        "type": "string",
                        "enum": ["Serum", "Plasma", "Heart Tissue", "Not Sure"],
                        "description": f"Was human {mirna_oi} measured in serum, plasma, heart tissue or somewhere else?",
                    },
                    # Q5
                    "mortality": {
                        "type": "string",
                        "enum": ["Yes", "No", "Not Sure"],
                        "description": f"Is {mirna_oi} in the abstract clearly associated with mortality/death?",
                    },
                    # Q6
                    "measurement_type": {
                        "type": "string",
                        "enum": ["qPCR", "NGS", "Microarray", "Combination", "Not Sure"],
                        "description": f"Was human {mirna_oi} measured by qPCR, sequenced (NGS), used microarray?",
                    },
                    # Q7
                    "sample_size": {
                        "type": "integer",
                        "description": "What was the human sample size of the study?"
                    }
                },
                "required": ["related_topic", "direction_upreg_downreg", "primary_literature",
                             "serum_plasma_tissue", "mortality", "measurement_type", "sample_size"],
            },
        }
    ]
    return functions


# function to run model
# https://www.mlq.ai/gpt-function-calling-getting-started/

def run_conversation(GPT_MODEL, system_msg, user_msg, custom_gpt_function, retries=5):
    '''
    :param GPT_MODEL:           gpt model name
    :param system_msg:          describe the behavior of the AI assistant
    :param user_msg:            input text (abstract in our case)
    :param custom_gpt_function: gpt function calling syntax (structured questions here)
    :param retries:             number of retries  # could also use retry from tenacity library with the @retry decorator
    :return:                    response from gpt model
    '''

    try:  # could alternatively use requests library to call the API
        response = openai.ChatCompletion.create(
            model=GPT_MODEL,
            max_tokens=2048,
            temperature=0.7,
            top_p=1,
            messages=[{"role": "system", "content": system_msg},
                      {"role": "user", "content": user_msg}],
            functions=custom_gpt_function,
            function_call="auto",
        )
        # message = response["choices"][0]["message"]
        return response

    except Exception as e:
        if retries > 0:
            print(f"An error occurred: {str(e)}. {retries} left - Retrying in 20 seconds ...")
            time.sleep(20)
            print(f"Retrying {retries} more times")
            return run_conversation(GPT_MODEL, system_msg, user_msg, custom_gpt_function, retries - 1)
        else:  # after trying {retries} times
            print(f"An error occurred: {str(e)}. No retries left.", file=sys.stderr)
            return None


# ==========================
# Preparation
# ==========================

# %% load data
path2dat = "./data-literature/pubmed/"
list_of_files = sorted(os.listdir(path2dat))
concatenated_files = [path2dat + file for file in list_of_files if "-2023-06-16-" in file]

diseases = ["ACS", "CAD", "DCM", "HFrEF"]

# create pandas dataframe with disease abbreviations and full names
diseases_df = pd.DataFrame({"abbrev_disease": diseases,
                            "disease_name": ["Acute Coronary Syndrome", "Coronary Artery Disease",
                                             "Dilatative Cardiomyopathy", "Heart Failure"]
                            })

# ==========================
# Loop over diseases
# ==========================
# %%  loop over diseases and abstracts
current_date = datetime.now().strftime("%Y-%m-%d")

for i, disease in enumerate(diseases_df.disease_name.tolist()[1:], start=1):
    print(f"Processing abstracts for {disease} ({i+1}/{diseases_df.shape[0]})...")
    # ==========================
    df = pd.read_csv(concatenated_files[i], sep=";", low_memory=False)
    # initialize empty col names
    new_columns = ['usage_total_tokens', 'query_cost', 'time']
    # Using assign() method to add empty columns
    df = df.assign(**{col: pd.Series(dtype='float64') for col in new_columns})
    # Columns to change dtype to "object"
    columns_to_change = ['related_topic', 'direction_upreg_downreg', 'primary_literature', 'serum_plasma_tissue', 'mortality', 'measurement_type']
    df[columns_to_change] = df[columns_to_change].astype('object')
    df = df.drop('unrelated_topic01', axis=1)
    # df['time'] = df['time'].astype(float)
    # df.head()
    # df.info(); df.shape
    my_abstracts_grouped = df.groupby('PMID').agg({'miRNA': lambda x: list(x.unique())}).reset_index()
    # ==========================
    # loop over abstracts
    for index, row in df.iterrows():
        print(f"GPT is processing abstract and miRNA {index+1}/{df.shape[0]}...")
        # ==========================
        custom_gpt_function = my_gpt_api_function(abstract_disease=disease, mirna_oi=df["miRNA"][index])
        system_msg = 'You are a helpful assistant who understands data science and cardiovascular molecular biology.'
        #break
        # ==========================
        start_time = time.time()
        # run model
        gpt_response = run_conversation(
            GPT_MODEL=GPT_MODEL,
            system_msg=system_msg,
            user_msg=df.Abstract[index],
            custom_gpt_function=custom_gpt_function,
            retries=5
        )
        elapsed_time = time.time() - start_time  # in seconds
        # clean response
        message_dict = json.loads(gpt_response.choices[0]["message"]["function_call"]["arguments"])
        # test1 = pd.DataFrame.from_records([message_dict])  # convert dict to df
        prompt_tokens = gpt_response.usage["prompt_tokens"]
        completion_tokens = gpt_response.usage["completion_tokens"]
        total_tokens = prompt_tokens + completion_tokens
        query_cost = prompt_tokens/1000*cost_per_1000_input_tokens + completion_tokens/1000*cost_per_1000_output_tokens
        # ==========================
        # fill up dataframe
        # mirna data
        df.loc[index, 'related_topic'] = message_dict["related_topic"]
        df.loc[index, 'direction_upreg_downreg'] = message_dict["direction_upreg_downreg"]
        df.loc[index, 'primary_literature'] = message_dict["primary_literature"]
        df.loc[index, 'serum_plasma_tissue'] = message_dict["serum_plasma_tissue"]
        df.loc[index, 'mortality'] = message_dict["mortality"]
        df.loc[index, 'measurement_type'] = message_dict["measurement_type"]
        df.loc[index, 'sample_size'] = message_dict["sample_size"]
        # usage
        df.loc[index, 'usage_total_tokens'] = total_tokens
        df.loc[index, 'query_cost'] = query_cost
        df.loc[index, 'time'] = elapsed_time
        # ==========================
        # store original openai class object
        file_name_object = f"./output/gpt_response/{diseases[i]}/{current_date}-index{index}-gpt_response.pkl"
        with open(file_name_object, 'wb') as f:
            pickle.dump(gpt_response, f)
        # ==========================

    # store filled dataframe
    file_name_df_csv = f"./output/gpt_dataframe/{diseases[i]}/{current_date}-df-pubmed-gpt_response.csv"
    file_name_df_pkl = f"./output/gpt_dataframe/{diseases[i]}/{current_date}-df-pubmed-gpt_response.pkl"
    file_name_df_xlsx = f"./output/gpt_dataframe/{diseases[i]}/{current_date}-df-pubmed-gpt_response.xlsx"
    df.to_csv(file_name_df_csv, index=False)
    df.to_pickle(file_name_df_pkl)
    df.to_excel(file_name_df_xlsx, index=False)
    print(f"Finished processing abstracts for {disease}...\nThank you for your patience.\n==========================\n")
    # ========================== END OF LOOP ========================== #

File: scripts/gpt/R/gpt-abstract-screen.R

# the query was made on personal Mac M1 with API key stored in environment vars

# CONSIDER
## Default rate limits  https://platform.openai.com/docs/guides/rate-limits
## https://github.com/openai/openai-cookbook/blob/main/examples/How_to_handle_rate_limits.ipynb
## 3,500 RPM (requests per minute)
## 90,000 TPM (tokens per minute)


# Get system name
system_name <- Sys.info()["nodename"]
mount_filesystem <- TRUE
saveFILE <- TRUE

# Define library and data paths based on system
if (system_name == "MacBook-Pro-CR-2065.local") {
  lib_path <- .libPaths()[1]
  data_path_bestageing2022 <- "/Volumes/T7CR/data/bestageing2022"
  data_path_BestAgeing <- "/Volumes/T7CR/data/BestAgeing"
  if(mount_filesystem == TRUE) {
    data_path_bestageing2022 <- "/Users/christophreich/Desktop/mount/rockerprojects/bestageing2022"  # mount -t nfs 10.55.1.185:/data/users/reich/ ~/Desktop/mount/
    data_path_BestAgeing <- "/Users/christophreich/Desktop/mount/BestAgeing"
  }
} else {  # assuming cluster
  lib_path <- "/mnt/users/reich/programs/R43/lib" 
  data_path_bestageing2022 <- "/mnt/users/reich/rockerprojects/bestageing2022"
  data_path_BestAgeing <- "/mnt/users/reich/BestAgeing"
}

#  Load libraries
require(miRetrieve, lib.loc = lib_path); packageVersion('miRetrieve')
require(glue, lib.loc = lib_path)
require(magrittr, lib.loc = lib_path) # Load magrittr for %>%
require(ggplot2, lib.loc = lib_path) # Load ggplot2 for plotting
require(svglite, lib.loc = lib_path)
require(easyPubMed, lib.loc = lib_path) # Load easyPubMed, access to PubMed
require(rcrossref, lib.loc = lib_path)
require(dplyr, lib.loc = lib_path) # Data wrangling
require(tidyr, lib.loc = lib_path) # Data wrangling
require(patchwork, lib.loc = lib_path) # Group graphs
#easyPubMed does not work
require(rentrez, lib.loc = lib_path)
require(pubmedR, lib.loc = lib_path)



library(openai)
library(chatgpt)
library(httr)
library(jsonlite)
library(reticulate)  # for 'tiktoken' library to count tokens
library(tidyverse)
library(glue)
## comment out if this script is used in the future!! 
# use_python("/Library/Frameworks/Python.framework/Versions/3.11/bin/python3", required = T)
# tiktoken <- import("tiktoken")
# encoding <- tiktoken$encoding_for_model("gpt-3.5-turbo")  # model
# myapikey <- Sys.getenv("OPENAI_API_KEY")
# gptmodel <- openai::list_models()$data$root[39]   # gpt-3.5-turbo-0613

source(file = "./scripts/gpt/gpt-fns.R")


# load data -----------------------------------------------------

diseases <- c("ACS", "CAD", "DCM", "HFrEF")
diseases_tibble <- tibble(abbrev_disease = diseases,
                          disease_name = c("Acute Coronary Syndrome", "Coronary Artery Disease", "Dilatative Cardiomyopathy", "Heart Failure"))
path2dat <- paste0(data_path_bestageing2022, "/data-literature/pubmed/", list.files(path = glue("{data_path_bestageing2022}/data-literature/pubmed/")))
path2dat <- path2dat[grep("-2023-06-16-", path2dat)]   # use newest files from 2023-06-16

cost_input <- numeric()
processed_input_tokens <- numeric()
for(disease in seq_along(diseases)) {
  
  
  my_abstracts <- as_tibble(read.csv2(file = path2dat[disease]))
  print(glue("For {diseases[disease]} {length(unique(my_abstracts$PMID))} unique abstracts were found."))
  print(glue("For {diseases[disease]} {length(unique(my_abstracts$miRNA))} unique miRNAs were found."))
  # check api costs
  #calculate_input_token_cost() # now done in helper-fn in "gpt-fns.R"
  
}

length(unique(my_abstracts$PMID))
length(unique(my_abstracts$miRNA))

grouped_abstracts <- my_abstracts %>% 
  group_by(PMID) %>% 
  summarize(miRNA = list(unique(miRNA)))

# initialize column where we save our query tibbles later
grouped_abstracts$df_temp_abstract_x <- vector("list", nrow(grouped_abstracts))
grouped_abstracts$answer <-  character(length = nrow(grouped_abstracts))
grouped_abstracts$usage_total_tokens <- integer(length = nrow(grouped_abstracts))
grouped_abstracts$query_cost <- numeric(length = nrow(grouped_abstracts))
grouped_abstracts$time <- numeric(length = nrow(grouped_abstracts))


# get disease name for query
abstract_disease <- diseases_tibble$disease_name[disease]
for (unique_abstract in 1:nrow(grouped_abstracts)) {
  # get abstract: CAVE dim(my_abstracts)[1] != dim(grouped_abstracts)[1], that's why we are using the first index with "min"
  selected_abstract_index <- min(which(my_abstracts$PMID %in% grouped_abstracts$PMID[unique_abstract]))
  selected_abstract <- my_abstracts$Abstract[selected_abstract_index]
  
  # get mirna set of interest
  abstract_mirna_list <- unlist(grouped_abstracts$miRNA[unique_abstract])
  abstract_mirna_list_prompt <- paste0(abstract_mirna_list, collapse = ", ")
  system_prompt <- glue('[no prose, no explanation, no notes] [output only json] You are a Data Scientist analyzing microRNA data. You read medical abstracts and return answers 
                        with the column names "related_topic" for Q1, "direction_upreg_downreg" for Q2, "primary_literature" for Q3, 
                        "serum_plasma_tissue" for Q4, "mortality" for Q5, "measurement_type" for Q6, and "sample_size" for Q7.
                        
                        You provide {length(abstract_mirna_list)} separate json structured data for each of the following {length(abstract_mirna_list)} microRNAs in this ordered list: MY_LIST=[{abstract_mirna_list_prompt}]
                        
                        Q1: Does the abstract deal with {abstract_disease} in humans? (Yes/No)
                        Q2: Is the given miRNA from MY_LIST up- (increased) or downregulated (decreased) in {abstract_disease}? (Upregulated/Downregulated)
                        Q3: Is the article an original research and not a secondary literature (review articles, systematic reviews, meta-analysis)? (Yes/No)
                        Q4: Was the miRNA from MY_LIST measured in serum, plasma, heart tissue or somewhere else? (Serum/Plasma/Tissue/Else)
                        Q5: Is miRNA from MY_LIST in the abstract clearly associated with mortality/death? (Yes, No, Not given)
                        Q6: How was  miRNA from MY_LIST measured?
                        Q7: What was the human sample size in the study? (number)')
  user_prompt <- glue('Investigate the following Abstract: "{selected_abstract}"')
  
  # call to API
  resultsFromQuery <- retry_chatGPT(system_prompt = system_prompt, user_prompt = user_prompt, retries = 3)
  
  # resultsFromQuery <- chatGPT(system_prompt = system_prompt, user_prompt = user_prompt, cost_per_1000_input_tokens = 0.0015,
  #                             modelName = "gpt-3.5-turbo", temperature = 0.7)
  ## store results
  #grouped_abstracts$df_temp_abstract_x[[unique_abstract]] <- resultsFromQuery$df_temp_abstract_x
  grouped_abstracts$usage_total_tokens[unique_abstract] <- resultsFromQuery$usage_total_tokens
  grouped_abstracts$query_cost[unique_abstract] <- resultsFromQuery$query_cost
  grouped_abstracts$time[unique_abstract] <- resultsFromQuery$time
  grouped_abstracts$answer[unique_abstract] <- resultsFromQuery$answer
  
  print(glue::glue("|||-----------------------Run finished for unique abstract: {unique_abstract}/{nrow(grouped_abstracts)}-----------------------|||"))
}
sum_cost <- glue("{round(sum(grouped_abstracts$query_cost),3)} $")

saveRDS(object = grouped_abstracts, file = glue("./data-literature/pubmed-gpt/raw_results/{diseases[disease]}-{Sys.Date()}-human-gptpubmed_query-unique-articles{nrow(grouped_abstracts)}.rds"))



# saveRDS(object = grouped_abstracts, file = glue("./data-literature/pubmed-gpt/query_results/{diseases[disease]}-{Sys.Date()}-human-gptpubmed_query-unique-articles{nrow(grouped_abstracts)}.rds"))
